src/platform/PICO/adc_pico.c

```diff
@@ -0,0 +1,260 @@
+
+    // Set channels (ioTags) for enabled ADC inputs
+    if (config->vbat.enabled) {
+        adcOperatingConfig[ADC_BATTERY].tag = config->vbat.ioTag;
+    }
+
+    if (config->rssi.enabled) {
+        adcOperatingConfig[ADC_RSSI].tag = config->rssi.ioTag;
+    }
+
+    if (config->external1.enabled) {
+        adcOperatingConfig[ADC_EXTERNAL1].tag = config->external1.ioTag;
+    }
+
+    if (config->current.enabled) {
+        adcOperatingConfig[ADC_CURRENT].tag = config->current.ioTag;
+    }
+
+    uint32_t mask = 0u;
+
+    // loop over all possible channels and build the adcOperatingConfig to represent
+    // the set of enabled channels
+    for (int i = 0; i < ADC_SOURCE_COUNT; i++) {
+        if (!adcOperatingConfig[i].tag || i > PICO_ADC_MAX_CHANNELS) {
+            continue;
+        }
+
+        const int pin = DEFIO_TAG_PIN(adcOperatingConfig[i].tag);
+        const int channel = adcChannelByPin(pin);
+        if (channel >= 0) {
+            adcOperatingConfig[i].channel = channel;
+            adc_gpio_init(pin);
+            adcOperatingConfig[i].enabled = true;
+        }
+        mask |= (1u << channel);
+    }
+
+    // Map each enabled source to its DMA slot by channel rank (ascending channel order).
+    for (int i = 0; i < ADC_SOURCE_COUNT; i++) {
+        if (!adcOperatingConfig[i].enabled) {
+            continue;
+        }
+        const uint8_t ch = adcOperatingConfig[i].channel;
+        adcOperatingConfig[i].dmaIndex = (uint8_t)popcount(mask & ((1u << ch) - 1u));
+    }
+
+    const uint8_t channelCount = popcount(mask);
+    if (channelCount == 0) {
+        /* don't enable the adc/dma - exit immediately */
+        return;
+    }
+
+    if (channelCount == 3) {
+        // padding required due to DMA ring buffer supporting 2^X
+        // enable the internal temp sensor to add one more reading.
+        adc_set_temp_sensor_enabled(true);
+        mask |= (1u << PICO_ADC_INTERNAL_TEMP_CHANNEL);
+    }
+
+    // empty the ADC fifo for safety
+    adc_fifo_drain();
+
+    // set the round robin for ADC
+    adc_set_round_robin(mask);
+
+    const int firstEnabled = llog2(mask & -mask);
+    adc_select_input(firstEnabled);
+
+    // Write each completed conversion to the sample FIFO
+    // Enable DMA data request (DREQ)
+    // Fire DREQ when FIFO has at least 1 sample
+    // Ignore ERR bit, save masking the result (as it is 12 bit).
+    // Don't shift each sample to 8 bits when pushing to FIFO
+    adc_fifo_setup(true, true, 1, false, false);
+
+    /*
+        Conversion requires 96 cycles per sample, at ADC clock of 48MHz.
+        clkdiv divides ADC clock to generate next auto trigger event
+
+        clkdiv = (48,000,000 / (sample_rate)) - 1
+
+        Sample rate required at 10hz for X channels = 10 x X
+
+        65535.f + 255.f / 256.f for as slow as possible
+        0 for as fast as possible
+    */
+    adc_set_clkdiv(65535.f + 255.f / 256.f);
```

Comment:
maybe just adc_set_clkdiv(65535); - it will sample every 65536 cycles. Using fractional part is .. strange (one of 256 conversion would be faster, 65537 is not power of 2)

But it will be better to set some predictable sampling frequency. Just handle padding correctly

Reply:
The fractional component was simply to make it as slow as possible... there is only a 1 byte available for the fraction. But I agree it makes it look more complicated for marginal benefit.